akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -2285,6 +2285,281 @@ private[akka] final class StatefulMap[S, In, Out](
+                // onComplete callback has run
+                case Supervision.Stop => failStage(ex)
+                case _                =>
+                  // try next element
+                  pushNextIfPossible()
+              }
+            case Failure(ex) =>
+              // fatal exception in buffer, not sure that it can actually happen, but for good measure
+              throw ex
+          }
+        }
+
+      private def pullIfNeeded(): Unit = {
+        // this will only be called after state future was completed
+        // so no need to check if state future is still pending or first element buffer is still full
+        if (isClosed(in) && buffer.isEmpty) releaseThenCompleteStage()
+        else if (earlyPulledElem.isEmpty && buffer.used < parallelism && !hasBeenPulled(in)) tryPull(in)
+        // else already pulled and waiting for next element
+      }
+
+      def initState(): Unit = {
+        val future = create()
+        stateAcquired = true
+        future.onComplete { s =>
+          initCB(s).failed.foreach {
+            case _: StreamDetachedException =>
+              // stream stopped before created callback could be invoked, we need
+              // to close the resource if it is was opened, to not leak it
+              s match {
+                case Success(r) =>
+                  // there is no way that out is still available
+                  onComplete(r)
+                // it's ok to not set the stateAcquired flag here
+                // since stream already been tear down anyway
+                case Failure(ex) =>
+                  throw ex // failed to open but stream is stopped already
+              }
+            case _ => // we don't care here
+          }(materializer.executionContext)
+        }(akka.dispatch.ExecutionContexts.parasitic)
+      }
+
+      private def resetState(): Unit = {
+        tryPushAfterInitialized = false
+        state match {
+          case Some(s) =>
+            val future = onComplete(s)
+            future.value match {
+              case Some(v) => emitThenFailOrCompleteStage(None, v); initState()
+              case None    => future.onComplete(v => releaseCB(None -> v))(akka.dispatch.ExecutionContexts.parasitic)
+            }
+          // unlikely to happen, but good for measurement anyway
+          case None => initState()
+        }
+      }
+
+      private def applyUserFunction(elem: In): Unit = {
+        try {
+          val future = f(state.get, elem)
+          val holder = new Holder[(S, Out)](NotYetThere, futureCB)
+          buffer.enqueue(holder)
+          future.value match {
+            case None =>
+              future.onComplete(holder)(akka.dispatch.ExecutionContexts.parasitic)
+            case Some(v) =>
+              // #20217 the future is already here, optimization: avoid scheduling it on the dispatcher and
+              // run the logic directly on this thread
```

Comment:
// #20217 the future is already here, optimization: avoid scheduling it on the dispatcher and
seems not needed